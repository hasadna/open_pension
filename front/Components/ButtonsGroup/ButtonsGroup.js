import {useState} from 'react';
import InfoTooltip from "../InfoTooltip/InfoTooltip";

export default function ButtonGroups({title, buttons, selectHandler, defaultActiveButton, description = null, multiple = false}) {

  const [activeButtons, setActiveButtons] = useState(() => {
    if (defaultActiveButton) {
      return {[defaultActiveButton]: true};
    }

    return {};
  });

  const existsSelectedButtons = Object.keys(activeButtons);
  const optionIsSelected = (identifier) => {
    if (existsSelectedButtons.includes(identifier)) {
      return activeButtons[identifier];
    }

    return false;
  }

  const handleButtonClick = (e) => {
      e.preventDefault();

      const {target: {dataset: {identifier}}} = e;

      if (multiple) {
        setActiveButtons({...activeButtons, ...{[identifier]: !optionIsSelected(identifier)}});
      }
      else {
        setActiveButtons({[identifier]: !optionIsSelected(identifier)});
      }

      if (selectHandler) {
        selectHandler(activeButtons);
      }
  };

  const getButtonClass = (identifier) => optionIsSelected(identifier) ? 'active' : null;

  return <div className="buttons-group">
    <div className="title-wrapper">
      <span className="title">{title}</span>
      {description && <div className="info-wrapper"><InfoTooltip description={description} /></div>}
    </div>

    <ul>
      {Object
        .entries(buttons)
        .map(([identifier, value], key) =>
          <li key={key}>
            <a href="#"
               onClick={handleButtonClick}
               data-identifier={identifier}
               className={getButtonClass(identifier)}
            >{value}</a>
          </li>
        )
      }
    </ul>
  </div>
}
