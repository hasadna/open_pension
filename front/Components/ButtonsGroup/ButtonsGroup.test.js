import React from 'react'
import {render, screen, fireEvent} from '@testing-library/react';
import renderer from 'react-test-renderer';
import './ButtonsGroup'
import ButtonGroups from "./ButtonsGroup";

let selectHandler = jest.fn();

describe('Buttons group', () => {

  const getComponent = (props = {}) => <ButtonGroups
    title={"Testing component"}
    description={'This is the description'}
    buttons={['First option', 'Second option']}
    selectHandler={selectHandler}
    {...props}
  />;

  const clickOnLink = (link, calledWith) => {
    // Click on the first element.
    fireEvent.click(link);
    expect(selectHandler).toBeCalledWith(calledWith);
  }

  it('Snapshot', () => {
    const trees = {
      full: getComponent(),
      noDescription: getComponent({description: ''})
    };

    Object.values(trees).forEach((treeToRender) => {
      const tree = renderer.create(treeToRender).toJSON()
      expect(tree).toMatchSnapshot();
    });
  });

  it('Testing the elements', () => {
    render(getComponent());
    expect(screen.getByTestId('title')).toBeInTheDocument();
    expect(screen.getByTestId('description')).toBeInTheDocument();
    expect(screen.getByTestId('items-wrapper')).toBeInTheDocument();
    expect(screen.getAllByTestId('item').length).toBe(2);
  });

  it('Testing when rendering without a description', () => {
    render(getComponent({description: ''}));
    expect(screen.queryByTestId('description')).not.toBeInTheDocument();
  });

  it('Testing a selection of a single item', () => {
    render(getComponent());
    expect(selectHandler).not.toBeCalled();

    const [firstLink, secondLink] = screen.getAllByTestId('item');

    expect(firstLink.text).toBe('First option');
    expect(secondLink.text).toBe('Second option');

    clickOnLink(firstLink, {"0": true});

    // Verifying classes.
    expect(firstLink).toHaveClass('active');
    expect(secondLink).not.toHaveClass('active');

    clickOnLink(secondLink, {"1": true})

    // Verifying classes.
    expect(secondLink).toHaveClass('active');
    expect(firstLink).not.toHaveClass('active');

    // Testing the total number invocations.
    expect(selectHandler).toBeCalledTimes(2);
  });

  it('Testing selection of multiple elements', () => {
    render(getComponent({multiple: true}));
    expect(selectHandler).not.toBeCalled();

    const [firstLink, secondLink] = screen.getAllByTestId('item');

    expect(firstLink.text).toBe('First option');
    expect(secondLink.text).toBe('Second option');

    // Turn on both of the items.
    clickOnLink(firstLink, {"0": true});
    expect(firstLink).toHaveClass('active');
    expect(secondLink).not.toHaveClass('active');

    clickOnLink(secondLink, {"0": true, "1": true});
    expect(secondLink).toHaveClass('active');
    expect(firstLink).toHaveClass('active');

    // Click on the first item again and verify the classes.
    clickOnLink(firstLink, {"0": true});
    expect(firstLink).not.toHaveClass('active');
    expect(secondLink).toHaveClass('active');

    // Testing the total number invocations.
    expect(selectHandler).toBeCalledTimes(3);
  });
});
