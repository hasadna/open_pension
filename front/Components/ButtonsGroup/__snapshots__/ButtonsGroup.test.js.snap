// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Buttons group Snapshot 1`] = `
<div
  className="buttons-group"
>
  <div
    className="title-wrapper"
  >
    <span
      className="title"
      data-testid="title"
    >
      Testing component
    </span>
    <div
      className="info-wrapper"
      data-testid="description"
    >
      <div
        className="info-tooltip off"
      >
        <div
          className="a-wrapper"
        >
          <a
            href="#"
            onClick={[Function]}
          >
            i
          </a>
        </div>
      </div>
    </div>
  </div>
  <ul
    data-testid="items-wrapper"
  >
    <li>
      <a
        className={null}
        data-identifier="0"
        data-testid="item"
        href="#"
        onClick={[Function]}
      >
        First option
      </a>
    </li>
    <li>
      <a
        className={null}
        data-identifier="1"
        data-testid="item"
        href="#"
        onClick={[Function]}
      >
        Second option
      </a>
    </li>
  </ul>
</div>
`;

exports[`Buttons group Snapshot 2`] = `
<div
  className="buttons-group"
>
  <div
    className="title-wrapper"
  >
    <span
      className="title"
      data-testid="title"
    >
      Testing component
    </span>
    
  </div>
  <ul
    data-testid="items-wrapper"
  >
    <li>
      <a
        className={null}
        data-identifier="0"
        data-testid="item"
        href="#"
        onClick={[Function]}
      >
        First option
      </a>
    </li>
    <li>
      <a
        className={null}
        data-identifier="1"
        data-testid="item"
        href="#"
        onClick={[Function]}
      >
        Second option
      </a>
    </li>
  </ul>
</div>
`;
